### DISCUSSION TITLE
Ways to put `turbo` commands into the `package.json` of packages

### DISCUSSION THREAD

**Discussion by serkodev** [vercel/turborepo#9321] [Ideas]
### Goals

Allow to put `turbo` commands into the `package.json` of packages

### Background

For users who have not installed `turbo` globally, they will need to run commands within sub-packages using `pnpm turbo <command>` instead of `turbo <command>`. If they only run `pnpm build` without explicitly using Turbo, some dependencies' build from workspace might not be updated.

However, if we include `turbo run <command>` inside `package.json`, as mentioned in the [documentation](https://turbo.build/repo/docs/crafting-your-repository/running-tasks#using-scripts-in-packagejson), this could result in recursive calls.

> You only want to write turbo commands in your root package.json. Writing turbo commands into the package.json of packages can lead to recursively calling turbo.

So I want to find a way to let user able to put `turbo` command in sub packages `scripts` and able to run it in both inside sub packages or from root package.

Related: #1232

### Solution 1: Check if Turbo is running recursively

I’m not sure if there’s currently a way to check if Turbo itself is running recursively, but if possible, it would be helpful to have a mechanism that skips repeated tasks when this occurs.

### Solution 2: Add a `--prepare` flag to the `run` command

Introduce a `--prepare` flag to the `run` command, which will execute the `dependsOn` task only, without running the command from the package itself.

Example:

Original:

```json
{
  "scripts": {
    "build": "next build"
  }
}
```

After:

```json
{
  "scripts": {
    "build": "turbo run build --prepare && next build"
  }
}
```

Now, if we run `pnpm build`, it will execute `turbo run build --prepare`, followed by `next build`. If the command is run from the root, it will automatically detect that it is already running inside a Turbo process and skip the prepare task.

**Comment by AlexAegis** [comment]
What I'm doing is all my tasks are postfixed with `_` so I can have a non postfixed version that invokes turbo

For example I have this in individual packages in my monorepo:

```json
{
	"scripts": {
		"test": "turbo run test_ --concurrency 16 --filter @org/package",
		"test_": "vitest --passWithNoTests --coverage --run",
	}
}
```

If I understood your problem correctly, this approach would eliminate your problem. Here's an example of this in one of my repos: https://github.com/AlexAegis/js-tooling I don't directly invoke turbo and I can invoke tasks in sub-packages too and have turbo execute all dependencies as expected.

**Reply by serkodev** [reply]
BTW, for the scripts inside sub-packages, you can use it without `--filter` because of [Automatic Package Scoping](https://turbo.build/repo/docs/crafting-your-repository/running-tasks#automatic-package-scoping).

```diff
{
  "name": "@org/package",
  "scripts": {
-    "test": "turbo run test_ --concurrency 16 --filter @org/package",
+    "test": "turbo run test_ --concurrency 16",
    "test_": "vitest --passWithNoTests --coverage --run",
  }
}
```

**Comment by serkodev** [comment]
@AlexAegis Thanks, seems this is a workaround for now.

**Comment by anthonyshew** [comment]
This doesn't answer directly to the intent of your post, but I did notice this and wanted to make sure you knew about it.
> Introduce a --prepare flag to the run command, which will execute the dependsOn task only

This is possible with `turbo run my-task --filter=my-package^...`. This combination of [filter microsyntaxes](https://turbo.build/repo/docs/reference/run#microsyntaxes-for-filtering) only runs the task for `my-packages` dependencies, as you're describing.

**Comment by Braden1996** [comment]
I've been using `"test": "yarn run -T turbo test-turbo --filter=${npm_package_name}",` as it works dynamically and can be copy & pasted / search & replaced across packages easily.
